apps/openmw/mwmechanics/pathfinding.cpp

```diff
@@ -291,29 +296,28 @@ namespace MWMechanics
         const MWWorld::CellStore* cell,
         bool allowShortcuts)
     {
-        if (mPath.size() < 2)
-        {
-            // if path has one point, then it's the destination.
-            // don't need to worry about bad path for this case
-            buildPath(startPoint, endPoint, cell, allowShortcuts);
-        }
-        else
+        buildPath(startPoint, endPoint, cell, allowShortcuts);
+        // Path building logic will set closest path grid point as first
+        // waypoint, but if actor is between the first two waypoints, then
+        // actor should head to second waypoint.
+        if (mPath.size() >= 2)
         {
-            const ESM::Pathgrid::Point oldStart(*getPath().begin());
-            buildPath(startPoint, endPoint, cell, allowShortcuts);
-            if (mPath.size() >= 2)
+            std::list<ESM::Pathgrid::Point>::iterator iter = mPath.begin();
+            const ESM::Pathgrid::Point &firstWayPoint(*iter);
+            if (isBetweenWaypoints(startPoint, firstWayPoint, *++iter))
             {
-                // if 2nd waypoint of new path == 1st waypoint of old, 
-                // delete 1st waypoint of new path.
-                std::list<ESM::Pathgrid::Point>::iterator iter = ++mPath.begin();
-                if (iter->mX == oldStart.mX
-                    && iter->mY == oldStart.mY
-                    && iter->mZ == oldStart.mZ)
-                {
-                    mPath.pop_front();
-                }
+                mPath.pop_front();
             }
         }
     }
 
+    bool PathFinder::isBetweenWaypoints(const ESM::Pathgrid::Point& pointToCheck,
+        const ESM::Pathgrid::Point& firstWayPoint,
+        const ESM::Pathgrid::Point& secondWayPoint)
+    {
+        float distBetweenPoints = sqrDistanceIgnoreZ(firstWayPoint, secondWayPoint);
+        float distTo2ndPoint = sqrDistanceIgnoreZ(pointToCheck, secondWayPoint);
+        return (distTo2ndPoint < distBetweenPoints);
+    }
```

Comment:
Seems problematic. I can easily imagine a path where the "between waypoint" check fails. With the path in the attached diagram, the new code would send the actor running against the wall.
![path](https://cloud.githubusercontent.com/assets/720642/9026136/f6920f06-3923-11e5-8aca-57c4ab3fc4dc.png)